state/state.go

```diff
@@ -26,3 +32,27 @@ type AccAddress = sdk.AccAddress
 
 // Int is an alias to the Int type from Cosmos-SDK.
 type Int = math.Int
+
+func (a *Address) UnmarshalJSON(data []byte) error {
```

Comment:
Any reason to keep this as method on ptr, while Marshal is method on val?

Reply:
Will test without

Reply:
@Wondertan it panics without the pointer val. The marshal having a value receiver has to do with BalanceForAddress returning the value instead of the pointer, but it works on both pointers and values. 

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x48 pc=0x104356c4c]

goroutine 1 [running]:
encoding/json.(*encodeState).marshal.func1()
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:326 +0x80
panic({0x1053af8a0, 0x106d1efc0})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/runtime/panic.go:884 +0x204
github.com/celestiaorg/celestia-node/state.Address.MarshalJSON(...)
	/Users/rford/Programming/Celestia/celestia-node/state/state.go:57
encoding/json.marshalerEncoder(0x14000187c80, {0x105628700?, 0x14001912df0?, 0x105764fe0?}, {0x0?, 0x87?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:477 +0xb8
encoding/json.condAddrEncoder.encode({0x10579c480?, 0x10579c4d0?}, 0x140018e1528?, {0x105628700?, 0x14001912df0?, 0x105628700?}, {0x38?, 0x15?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:960 +0x78
encoding/json.(*encodeState).reflectValue(0x105340ce0?, {0x105628700?, 0x14001912df0?, 0x102b37198?}, {0xd8?, 0x15?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:358 +0x70
encoding/json.interfaceEncoder(0x14000187c80, {0x105340ce0?, 0x14001912de0?, 0x102907618?}, {0xa5?, 0xfd?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:714 +0xcc
encoding/json.arrayEncoder.encode({0x140018e1708?}, 0x14000187c80, {0x10527e020?, 0x14000018c28?, 0x80?}, {0xe0?, 0xed?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:914 +0xc0
encoding/json.sliceEncoder.encode({0x140018e17e8?}, 0x14000187c80, {0x10527e020?, 0x14000018c28?, 0x6?}, {0x9?, 0x0?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:887 +0x260
encoding/json.structEncoder.encode({{{0x14001618fc0?, 0x1057fc360?, 0x1400136c030?}, 0x140011d23c0?}}, 0x14000187c80, {0x1055a1740?, 0x14000018c00?, 0x1055a1740?}, {0x30?, 0x23?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:759 +0x178
encoding/json.(*encodeState).reflectValue(0x14000187c80?, {0x1055a1740?, 0x14000018c00?, 0x102b33298?}, {0x78?, 0x0?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:358 +0x70
encoding/json.(*encodeState).marshal(0x140018e1a58?, {0x1055a1740?, 0x14000018c00?}, {0x40?, 0x0?})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:330 +0x120
encoding/json.Marshal({0x1055a1740, 0x14000018c00})
	/opt/homebrew/Cellar/go/1.20.3/libexec/src/encoding/json/encode.go:161 +0x94
main.sendJSONRPCRequest({0x16d5073ed?, 0x11?}, {0x16d5073f3, 0x11}, {0x14001912de0, 0x1, 0x1})
	/Users/rford/Programming/Celestia/celestia-node/cmd/celestia/rpc.go:337 +0x118
main.glob..func5(0x106d401a0?, {0x14000018b80?, 0x4?, 0x4?})
	/Users/rford/Programming/Celestia/celestia-node/cmd/celestia/rpc.go:99 +0x90
github.com/spf13/cobra.(*Command).execute(0x106d401a0, {0x14000018b00, 0x4, 0x4})
	/Users/rford/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:920 +0x5b0
github.com/spf13/cobra.(*Command).ExecuteC(0x106d3fec0)
	/Users/rford/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x35c
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/rford/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/Users/rford/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:961
main.run()
	/Users/rford/Programming/Celestia/celestia-node/cmd/celestia/main.go:28 +0x64
main.main()
	/Users/rford/Programming/Celestia/celestia-node/cmd/celestia/main.go:21 +0x1c
```

Reply:
My point is mostly why this is asymmetric. I am fine with either, as long as it symmetric

Reply:
No, it cannot be symmetric - thats my point. The receiver needs to be a value type on the unmarshal